(feat): add methods for changing settings

ilan-gold opened the pull request:
<!-- Please:
1. Fill in the following check boxes
2. Make sure checks pass (Ignore “Triage” ones)
-->

- [x] Towards #890 - solving the issue will simply require calling `settings.register` for the needed variable
- [x] Tests added
- [x] Release note added (or unnecessary)

ilan-gold commented:
I am aware this was not your top choice @flying-sheep but you also didn't seem so jazzed about scanpy's solution either.  So I finished this up. If we want something different, we can go for it.  I think the best option might be something shared between scanpy and anndata.  What do you think?  Short of that, despite the unpythonic nature of this, it is quite simple IMO which is a big plus.

ilan-gold commented:
@flying-sheep I'm fine with imitating scanpy.  You sounded fairly negative about it, so I guess it didn't seem like it mattered if we used one "bad" solution or another.  Do you think this should be built with scanpy in mind i.e., migrating scanpy to whatever we do here (in some capacity)?

flying-sheep reviewed (changes requested):
> I'm fine with imitating scanpy. You sounded fairly negative about it, so I guess it didn't seem like it mattered if we used one "bad" solution or another.

Scanpy’s solution is … fine. The general idea is great: a perfectly Pythonic API; leniently allow setting `settings.verbosity = 'info'` and then the user can read out `setting.verbosity is Verbosity.info`.

It was made in a time when module-wide `__getattr__`/`__setattr__` wasn’t a thing, that’s why we went with properties and a singleton class. It doesn’t really matter how to implement it, as long as the API and documentation are nice and clean. For a highly user-facing API like this, complexity is very much secondary, so let’s go with the solution that has the least tradeoffs in documentation and API.

I was unenthused about the `override` mechanism (only for `verbosity`) and the `reset` mechanism (absent). Any solution with `get_` and `set_` can only ever feel out of place in a Python project.

> Do you think this should be built with scanpy in mind i.e., migrating scanpy to whatever we do here (in some capacity)?

Yeah, I think making this reusable would be nice if it becomes sufficiently complex.

ilan-gold commented on docs/release-notes/0.10.4.md:
````diff
@@ -14,3 +14,7 @@
 
 ```{rubric} Performance
 ```
+
+```{rubric} Other updates
+```
+* Add `describe_option`, `get_option`, `reset_option`, and `set_option` methods for altering internally-used options, like hcecking for uniqueness on `obs`' index {pr}`1270` {user}`ilan-gold`
````
Got it.

flying-sheep commented on anndata/_config.py:
```diff
@@ -0,0 +1,240 @@
+        self._config[option] = self._registered_options[option].default_value
+
+    @contextmanager
+    def override(self, **overrides):
```
This doesn’t support autocompletion, how can we fix that?

ilan-gold replied:
Ok so I have spent quite a long time on this to no avail.  I have tried many different things.  I will need to look at this again on Monday.  If this is a breaking point, we may have to go the scanpy route (i.e., explicit attributes/keyword/positional arguments).  Something I realized while working through this, though, is that this solution _is generalizable_ in that we could migrate scanpy towards this.  But with the positional/keyword args way of doing things (i.e., what scanpy does), I don't think we can really support inheritance/re-usability.  But maybe I am missing something - I think there are ways of passing documentation up an inheritance tree, but I am not sure about types.

flying-sheep replied:
you managed to do that now with `update_override_function`, no?

ilan-gold replied:
This works for non-static analyzers, like IPython, but not vscode, which uses static analysis.

flying-sheep commented on anndata/_config.py:
```diff
@@ -0,0 +1,240 @@
+
+    Parameters
+    ----------
+    key : str
+        The environment variable name.
+    default_value : str
+        The default value for `os.environ.get`.
+    allowed_values : Sequence[str] | None, optional
+        Allowable string values., by default None
+    cast : _type_, optional
+        Casting from the string to a (potentially different) python object, by default lambdax:x
```
```suggestion
    key
        The environment variable name.
    default_value
        The default value for `os.environ.get`.
    allowed_values
        Allowable string values., by default None
    cast
        Casting from the string to a (potentially different) python object, by default lambdax:x
```

flying-sheep commented on anndata/tests/test_config.py:
```diff
@@ -0,0 +1,91 @@
+    version = "0.1.0"
+    settings.deprecate(test_option, version, warning)
+    described_option = settings.describe(print_description=False).split("\n")
+    assert (
+        len(described_option) == 3
+    )  # first line is message, second two from deprecation
```
```suggestion
    # first line is message, second two from deprecation
    assert len(described_option) == 3
```

flying-sheep commented on anndata/tests/test_config.py:
```diff
@@ -0,0 +1,91 @@
+    version = "0.1.0"
+    settings.deprecate(test_option, version)
+    described_option = settings.describe(print_description=False).split("\n")
+    assert (
+        len(described_option) == 2
+    )  # first line is message, second from deprecation version
```
```suggestion
    # first line is message, second from deprecation version
    assert len(described_option) == 2
```

flying-sheep commented on anndata/tests/test_config.py:
```diff
@@ -0,0 +1,91 @@
+    assert settings[test_option] == default_val
+
+
+def test_deprecation():
```
Should accessing that option throw a DeprecationWarning or FutureWarning?

If so, you can maybe use https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.deprecated_call

flying-sheep reviewed (changes requested):
This looks much nicer! I think options should definitely have autocompletion, so please check if that works in Jupyter Lab.

I think we can keep the nice `register` API as long as runtime autocompletion is great.

ilan-gold commented on anndata/_config.py:
```diff
@@ -0,0 +1,240 @@
+    removal_version: str | None
+
+
+# TODO: inherit from Generic[T] as well after python 3.9 is no longer supported
+class RegisteredOption(NamedTuple):
+    option: str
+    default_value: object
```
We can't do `NamedTuple` and `Generic[T]` in python 3.9.  So if we did only a `Generic`, we'd have to implement an `__init__` method.  Not the worst thing in the world. But you're right about `default_val`

ilan-gold commented:
Wow ok @flying-sheep, I finally did it.  `__signature__` was the key - `__annotations__` did not really help with the tab completion but `__signature__` did (and you can see that even that was not so straightforward).  If you want to test out tab completion here is something:

```python
import anndata as ad

test_option_doc = "doc string!"
test_option_env_var = "ANNDATA_TEST_VAR"
test_option = "test_var"
default_val = False
test_doc = f"""\
{test_option}: bool
    My doc string!
"""

test_option_doc_2 = "doc string 2!"
test_option_env_var_2 = "ANNDATA_TEST_VAR 2"
test_option_2 = "test_var_2"
default_val_2 = False
test_doc_2 = f"""\
{test_option_2}: bool
    My doc string 2!
"""


def validate_bool(val, option):
    assert val in [True, False], f"{val} not valid boolean for option {option}."


ad.settings.register(
    test_option, default_val, test_doc, lambda v: validate_bool(v, test_option)
)

ad.settings.register(
    test_option_2, default_val_2, test_doc_2, lambda v: validate_bool(v, test_option_2)
)
```

Have a look!  Now that I'm writing this, I would like to get the docstring working as well.  Will do.

ilan-gold commented:
From meeting: make some sort of `freeze` method so that downstream users can't edit the underlying object.

flying-sheep commented on anndata/_config.py:
```diff
@@ -0,0 +1,332 @@
+
+    Returns
+    -------
+    Callable
+        The wrapped `SettingsManager.register` function.
```
Please put types into the annotations, not into docstrings.

```suggestion
    The wrapped `SettingsManager.register` function.
```

ilan-gold replied:
Missed that one and a few others when cleaning up.  Will fix them all.

flying-sheep commented on anndata/_config.py:
```diff
@@ -0,0 +1,332 @@
+        The wrapped `SettingsManager.register` function.
+    """
+
+    @wraps(register)
+    def _impl(
```
Why is `update_override_function` a decorator? Wouldn’t it be simpler if it was a function that’s called in `register`?

ilan-gold replied:
From previous iterations.  Refactored.

flying-sheep commented on anndata/_config.py:
```diff
@@ -0,0 +1,332 @@
+        self.override.__annotations__[option] = type(default_value)
+        # __signature__ needs to be updated for tab autocompletion in IPython.
+        # See https://github.com/ipython/ipython/issues/11624 for inspiration.
+        setattr(
+            self.override.__func__,  # Attributes of a method are immutable, so need to use __func__.
+            "__signature__",
+            signature(self.override).replace(
```
`setattr` is when the `name` parameter isn’t fixed

```suggestion
        # Attributes of a method are immutable, so need to use __func__.
        self.override.__func__.__signature__= signature(self.override).replace(
```

ilan-gold replied:
Huh, thought I had read this doesn't work because of `__signature__` needing to be initialized specially, but seems to work now.  Great!

flying-sheep replied:
`x.a = v` is exactly the same as `setattr(x, 'a', v)`.

Both go through `__setattr__` if available and so on.

flying-sheep commented on anndata/_config.py:
```diff
@@ -0,0 +1,332 @@
+            ),
+        )
+        # Update docstring for `SettingsManager.override` as well.
+        insert_index = self.override.__doc__.find("\n        Yields")
```
how does that behave if you register more than one option?

ilan-gold replied:
What do you mean?  This looks for the spot right before "Yields" which is where you'd want to insert a new option (i.e., end of the list after "Parameters").  It should work, I checked it on the build with some test options.

flying-sheep replied:
This code gets called every time a setting is registered.

And every time, `options` gets inserted after `Yields`. Won’t that insert all options multiple times?

ilan-gold replied:
The next line only adds the new option.  Will relabel the variable.

ilan-gold replied:
![Screenshot 2024-01-25 at 11 02 03](https://github.com/scverse/anndata/assets/43999641/61da496f-7fd4-4b9f-a3a1-3fbec7f53520)

flying-sheep commented on anndata/_config.py:
```diff
@@ -0,0 +1,332 @@
+        if hasattr(super(), option):
+            super().__setattr__(option, val)
+        elif option not in self._registered_options:
+            raise KeyError(
```
shouldn’t that be `AttributeError`?

flying-sheep commented on anndata/_config.py:
```diff
@@ -0,0 +1,332 @@
+            registered_option.validator(val)
+            self._config[option] = val
+
+    def __getattr__(self, option: str) -> object:
```
I guess that’s the only point where missing static typing is painful.

Using `settings.foo` anywhere in the codebase will only be `object`

ilan-gold replied:
So yes, I think you're right.  I had not checked in VSCode for the attribute completion (only IPython kernels).  VSCode's type checker is static so this is not going to happen.  This seems like a deal breaker if I am not mistaken?

flying-sheep replied:
> This seems like a deal breaker if I am not mistaken?

depends on if we expect the settings mechanism to be used from library code or almost exclusively from interactive code.

flying-sheep commented on anndata/_config.py:
```diff
@@ -0,0 +1,332 @@
+##################################################################################
+
+
+def check_and_get_environ_var(
```
what’s that for? I don’t see it being used in the other code

ilan-gold replied:
This was in advance of using environment variables (i.e., for retaining categories).  I can just move it to the next PR.  It's tested and I was using for prototyping, that's why I did it here.

ilan-gold replied:
Removed.

flying-sheep reviewed (commented):
looks mostly good!

I’m still not sure if the registration way (and it’s lack of static type information) is preferable to simple properties combined with a unit test that checks that `override` and `reset` match up.

I’ll try it out a bit in an interactive session and see how I feel about it.

ilan-gold commented:
List of features that might be enabled by this branch:

- https://github.com/scverse/anndata/issues/610
- https://github.com/scverse/anndata/issues/890
- https://github.com/scverse/anndata/issues/1112

I think these are all pretty advanced use-cases.

Other things I could see us moving into some sort of settings:
- If/when we transition to `xarray`, assuming we stick with the same `AnnData` object, a variable for using the "experimental" `xarray` implementation (instead of releasing via `experimental`)
- io engines (rust or tensorstore or the like for zarr?)
- mutability behavior (see https://github.com/scverse/anndata/issues/742)

Honestly, though, these first two potential use-cases should probably not be set via an environment variable and is far as the third should, it would be under the "only if you know what you're doing" banner.  But not sure it's an env variable situation anyway.

At the end of the day, most of pandas and xarray's settings are about display and/or dependencies and we just don't do either too much at the moment (and in so far as we do external dependencies, this has been handled differently).  Furthermore, display only really matters in IPython settings, where autocomplete is present as things stand.  So I guess I really only see this as being "do this if you know what you're doing" as far as VSCode audience goes.  I'd be fine leaving things as is especially given the numerous advantages of the current approach (it's extendable, has deprecation warnings, and is very clean).

ivirshup commented:
> List of features that might be enabled by this branch:

I would also include the path towards:

* https://github.com/scverse/anndata/issues/746

For controlling if we auto convert to these, which we should do eventually.

> pandas and xarray's settings are about display and/or dependencies and we just don't do either too much at the moment 

Pandas does do feature flags for upcoming behaviour with settings flags. For example, [copy on write](https://pandas.pydata.org/docs/whatsnew/v2.2.0.html#copy-on-write).

ilan-gold commented:
I would still describe these as "advanced use-cases."  We don't have arguments like `N_PCS` from scanpy or the like which are important and could be oft used.

flying-sheep commented:
In that case, I fully agree with

> So I guess I really only see this as being "do this if you know what you're doing" as far as VSCode audience goes.

and like the current design. Maybe we should document that it’s only for advanced use cases.

ilan-gold commented:
@flying-sheep @ivirshup I addressed your comments and cleaned up the docs a bit.  Please see https://github.com/scverse/anndata/pull/1340 for the first "implementation" if it will help with the review here.  I am happy with the state of things and am ready to merge.

flying-sheep commented on anndata/_config.py:
```diff
@@ -0,0 +1,290 @@
+        # Update docstring for `SettingsManager.override` as well.
+        insert_index = self.override.__doc__.find("\n        Yields")
+        option_docstring = "\t" + "\t".join(
+            self.describe(option, print_description=False).splitlines(True)
```
```suggestion
            self.describe(option, print_description=False).splitlines(keepends=True)
```

flying-sheep commented on anndata/_config.py:
```diff
@@ -0,0 +1,290 @@
+        raise AttributeError(f"{option} not found.")
+
+    def __dir__(self) -> Iterable[str]:
+        return sorted(super().__dir__() + list(self._config.keys()))
```
```suggestion
        return sorted((*dir(super()), *self._config.keys()))
```

flying-sheep commented on anndata/_config.py:
```diff
@@ -0,0 +1,290 @@
+        option
+            Optional override for the option type to be displayed.  Otherwise `type(default_value)`.
+        """
+        validator(default_value)
```
How about you just do something like this?

```suggestion
        try:
            validator(default_value)
        except (ValueError, TypeError) as e:
            e.add_note(f"in option {option}")
```

Then you can just do `validator=bool` instead of having to craft the message in each validator function

PS: we have some helper to add to `__notes__` in Python 3.9 I think

flying-sheep reviewed (approved):
Looks great. I can think of one simplification to the API, otherwise, this is done as far as I’m concerned.
